cmd/beaker/session.go

```
@@ -125,14 +125,12 @@ To pass flags, use "--" e.g. "create -- ls -l"`,
 		home := runtime.Mount{HostPath: u.HomeDir, ContainerPath: u.HomeDir}
 
 		// Mount in a Beaker-managed home directory by default, if there's one configured.
-		if !localHome {
-			if config, err := getExecutorConfig(); err == nil {
-				// TODO: u.Username is highly dependent on host configuration. We
-				// should consider using the stable Beaker user ID instead.
-				home.HostPath = filepath.Join(config.SessionHome, u.Username)
-				if err := os.MkdirAll(home.HostPath, 0700); err != nil {
-					return fmt.Errorf("couldn't create home directory: %w", err)
-				}
+		if config, err := getExecutorConfig(); err == nil && config.SessionHome != "" && !localHome {
```

We should probably return an error if `localHome` is true but `config.SessionHome` is `""`. As if that happens the user wants a non-local home directory, but we don't know where to put it...

```
if !localHome {
    if config, err := getExecutorConfig(); err == nil {
        if config.SessionHome = "" {
             return errors.New("unable to create virtual home directory as there's no configured root")
        }
    }
}
```

---

Please reword the proposed error message, it's not great :).

---

I'm confused, I thought `--local-home` meant use `/home/<user>`, which doesn't require `config.SessionHome` to be configured. We don't have a `--alt-home-dir` flag anymore. 

Right now, Beaker will use `/home` by default or whatever `config.SessionHome` is set to. If `--local-home` is passed Beaker will use `/home` instead of `config.SessionHome`.

---

Yup, but here we're considering the case where we're not supposed to use the `localHome`, right?

So if we're supposed to use a virtual home directory but the executor isn't configured to put that somewhere, we're in an errant case, right?

---

If `localHome` is false, then the `--local-home` flag wasn't passed. I thought that meant "use whatever the executor is configured to use" not "use a virtual home directory". Is that correct? I may be missing some context here since it seems like the handling of home directories went through a lot of change while I was gone.

---

Yup, that's right. This is what I'm calling a "virtual home directory" -- a home directory that's placed by the Beaker Executor and probably isn't the canonical home directory on the host.

The error case I think we need to handle is when we're told to use the `config.SessionHome` but can't because it's `""`. I think the behavior as-is is that the user's canonical home directory would be used instead, but I think it'd be better to throw an exception so that the user isn't surprised when things they write interactively end up in `/home/$(whoami)`.

---

I thought the desired end state was to remove the `--local-home` flag and always use `/home/$(whoami)` once all machines have sufficient space in `/home`. At least that's what I gathered from talking to Colin here: https://allenai.slack.com/archives/G01Q41VK3EY/p1624386357375100

I really don't want to output an error in this case because it essentially makes the `--local-home` flag required. One of our goals was to make sessions as easy to create as possible. Currently that's just `beaker session create`. A required flag would add a stumbling block there.

---

> I thought the desired end state was to remove the --local-home flag and always use /home/$(whoami) once all machines have sufficient space in /home

Yup, that's true.  

> I really don't want to output an error in this case because it essentially makes the --local-home flag required.

That makes sense. How about we make `config.SessionHome` required? The executor could fail with an error at startup if `config.SessionHome` isn't set. We should probably also validate that the path exists.

---

> Do we know what happens now if config.SessionHome is "" with your changes? Where will non-local home directories be written? The working directory of the beaker process?

Per our conversation, it'll fallback to the canonical home directory (i.e. `/home/$(whoami)` on our servers).